### PR TITLE
Spgemm hash promotion

### PR DESCRIPTION
**Pull request by mndevec**
Updating SpGEMM methods. These will be in the next KokkosKernels promotion to Trilinos.

1- New parameter selection on GPUs. This adjust the shared memory size, block size, vectorsize.
2- New meta-algorithm called SPGEMM_KK. This is the default algorithm from now on. ---> SPGEMM_KK_MEMORY = KKMEM is outdated. 
    It chooses the best algorithm to run based on the input structure. 
    For example, it may choose dense accumulators or sparse on CPUs. 
    On GPUs, it adjusts the old KKMEM parameters as above, or it may choose to run a block-based algorithm rather than a warp-based one (KKMEM).
3- Minor updates: changing the test driver calls. It now requires "--" before parameter name that was missing previously. 

4- The new parameters name is SPGEMM_KK. In the xml file, any of the below will run this default algorithm. Please use the ones below after the new kokkoskernels promotion:
    - "SPGEMM_DEFAULT",  "SPGEMM_KK", "SPGEMM_KK_MEMSPEED"

5- I will put benchmark results on the wiki page soon.

### CODE REVIEW

**Comment by mndevec** [comment]
I think Tpetra was setting the parameter to "SPGEMM_KK_MEMSPEED". It should automatically be handled. 
@csiefer2
@jhux2

**Comment by mndevec** [comment]
This must be thoroughly tested, but lets wait after the shutdown week to merge this.

File: perf_test/common/KokkosKernels_MatrixConverter.cpp [hunk 1]
```diff
@@ -86,13 +85,13 @@ int main (int argc, char* argv[]){
...
     std::cerr << "Usage:" << argv[0]
-              << " in_mtx matrixfile out_mtx output_file [symmetrize] [remove_diagonal] [transpose]" << std::endl;
+              << " --in_mtx matrixfile --out_mtx output_file [--symmetrize] [--remove_diagonal] [--transpose]" << std::endl;
     std::cerr << "Input format .mtx for matrix market, .bin for binary, .crs for crs format" << std::endl;
     std::cerr << "Output format .mtx for matrix market, .bin for binary, .crs for crs format, .ligra for ligra output format" << std::endl;
 
     exit(1);
   }
-  typedef Kokkos::DefaultExecutionSpace MyExecSpace;
+  typedef Kokkos::OpenMP MyExecSpace;
```
[the remark this answers is not in the document]
Comment (mndevec) [reply perf_test/common/KokkosKernels_MatrixConverter.cpp]: Thanks for catching this. This is a utility executable to change input output formats. I will use your fix.

**Comment by mndevec** [comment]
@mhoemmen 
I did some updates based on your comments. 
Yes, this is a big commit to go over, I don't think it is possible to review all.
 
The main code is at KokkosSparse_spgemm_impl_kkmem.hpp from line-610 to 900. But that would  require algorithmic review.

**Comment by csiefer2** [comment]
Tpetra uses SPGEMM_KK_MEMORY by default.

File: perf_test/graph/KokkosGraph_triangle.cpp [hunk 2]
```diff
@@ -53,70 +53,70 @@
...
-  std::cerr << "chunksize [chunksize]              : how many vertices are executed with in a loop index. Default is 16." << std::endl;
-  std::cerr << "sort_option [0|1|2]                : How lower triangle will be sorted. 0: for largest to bottom, 1 for largest to top, 2 for interleaved." << std::endl;
-  std::cerr << "cache_flush [0|1|2]                : Flush between repetitions. 0 - no flush, 1 - soft flush, 2 - hard flush with random numbers." << std::endl;
-
-  std::cerr << "\nSuggested use of LL: executable amtx path_to_file.bin algorithm TRIANGLELL repeat 6 verbose chunksize [4|16]" << std::endl;
-  std::cerr << "Suggested use of LU: executable amtx path_to_file.bin algorithm TRIANGLELU repeat 6 verbose chunksize [4|16]" << std::endl;
-  std::cerr << "Suggested use of AI: executable amtx path_to_file.bin algorithm TRIANGLEIA repeat 6 verbose chunksize [4|16] rlt" << std::endl;
+  std::cerr << "--FLOP                               : Calculate and print the number of operations. This will be calculated on the first run." << std::endl;
+  std::cerr << "--COMPRESION [0|1]                   : Enable disable compression. Default:1." << std::endl;
```
Comment (mhoemmen) [inline perf_test/graph/KokkosGraph_triangle.cpp]: "COMPRESSION" (spelling)

**Comment by mhoemmen** [comment]
Thanks @mndevec !

**Pull request merged**